src/Rules/Drupal/Coder/FormNoPrivatePropertiesRule.php

```diff
@@ -0,0 +1,44 @@
+<?php declare(strict_types=1);
+
+namespace mglaman\PHPStanDrupal\Rules\Drupal\Coder;
+
+use Drupal\Core\Form\FormInterface;
+use PhpParser\Node;
+use PHPStan\Analyser\Scope;
+use PHPStan\Node\ClassPropertyNode;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+
+final class FormNoPrivatePropertiesRule implements Rule
+{
+
+    public function getNodeType(): string
+    {
+        return ClassPropertyNode::class;
+    }
+
+    public function processNode(Node $node, Scope $scope): array
+    {
+        if (!$node instanceof ClassPropertyNode
+            || !$node->getClassReflection()->implementsInterface(FormInterface::class)) {
+            return [];
+        }
+
+        if ($node->isPrivate()) {
+            return [
+                RuleErrorBuilder::message(
+                    'Private properties are note allowed on FormInterface classes due to seralisation.'
+                )->tip('See https://www.drupal.org/node/3110266')->build(),
+            ];
+        }
+        if ($node->isReadOnly()) {
+            return [
+                RuleErrorBuilder::message(
+                    'Private properties are note allowed on FormInterface classes due to seralisation.'
+                )->tip('See https://www.drupal.org/node/3110266')->build(),
+            ];
+        }
```

Comment:
Is this true? I thought read-only was only violated after being set. Or is it technically because it is set before __wakeup is called? Also, seems like a great candidate for its own rule class.

Reply:
Yes, it has to do the way Drupal serializes the class.
`Cannot initialize readonly property CLASS::readonlyproperty from scope Drupal\Core\Form\FormBase`
from `/core/lib/Drupal/Core/DependencyInjection/DependencySerializationTrait.php:89`
![image](https://github.com/mglaman/phpstan-drupal/assets/1375359/6f60a218-7d0d-46ba-818c-08d6acb5b8db)

And since it is triggering in both cases an invalid form, I suggest it shoud be on the same check.
check for valid form-classes for the serializer / serializable objects as a rule.

Reply:
That's fair! Can we have the rule return both errors then if they both exist instead of failing first on private and possibly failing again on readonly

PHPStan allows rules to return multiple errors